### PR TITLE
[Instrumentation.ServiceFabricRemoting] initial implementation

### PR DESCRIPTION
**Pull request by sablancoleis**
Today, Microsoft Service Fabric doesn't support trace context propagation natively. This library fills that gap and allows the trace context and Baggage to be transferred transparently between Services and Actors.

Note: This code targets only Service Fabric Remoting V2, since V1 would require using wrapped messages and it will be deprecated in the near future.

This implementation propagates the ActivityContext and Baggage through SF Remoting using the OpenTelemetry's Propagator infrastructure to inject the data into the message headers from the client and to extract the data from the headers on the receiving listener.

The code also provides extension methods to configure the whole infrastructure with one line of code in Program.cs:
tracerProviderBuilder.AddServiceFabricRemotingInstrumentation()

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by rajkumar-rangaraj** [comment]
@sablanc-msft Could you please sign the Easy CLA, it's a requirement to review/accept PRs.

**Comment by rajkumar-rangaraj** [comment]
@sablanc-msft You need to update the https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/.github/component_owners.yml file with owners. It is recommended to have a minimum of 2 owners. Once the component owner review is complete, approvers/maintainers will review.

Also, it would be nice if you provide a clear PR description of what you are trying to solve by adding this project to the repo.

**Comment by Kielek** [comment]
One more comment, do you see any possibility to include native-instrumentation directly in ServiceFabric? Whole Azure SDK is following this pattern, and it is the best option we can recommend.

**Comment by sablanc-msft** [comment]
> One more comment, do you see any possibility to include native-instrumentation directly in ServiceFabric? Whole Azure SDK is following this pattern, and it is the best option we can recommend.

That would be ideal. Hopefully the Service Fabric team can include native instrumentation in future releases. This library is meant to fill the gap of context propagation in the meantime.

**Comment by lmolkova** [comment]
@sablancoleis have you tried reaching out to the ServiceFabric team and proposing to add instrumentation there? 

It  seems they were considering adding instrumentation at some point and might be open to someone contributing it - https://github.com/microsoft/service-fabric-services-and-actors-dotnet/pull/294

**Comment by sablancoleis** [comment]
> @sablancoleis have you tried reaching out to the ServiceFabric team and proposing to add instrumentation there?
> 
> It seems they were considering adding instrumentation at some point and might be open to someone contributing it - [microsoft/service-fabric-services-and-actors-dotnet#294](https://github.com/microsoft/service-fabric-services-and-actors-dotnet/pull/294)

@lmolkova , Yes, I contacted them a month and a half ago and multiple times during the past few weeks, but I haven't been able to get a response yet. I will add you to the email thread.

I hope the SF Team integrates context propagation in a future release, but it will take months and even after that, it will take months for that release to be upgraded in the applications trying to use OpenTelemetry.
This library allows my Team, other teams in Microsoft and external customers using Service Fabric to integrate OpenTelemetry without having to upgrade the Service Fabric SDK, so I think it still adds value while we wait for the native integration.
My team is already using this code in production, but I would like to get it reviewed by the OpenTelemetry community and the Service Fabric team.

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs
Comment (olegsych) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs:29]: Is it possible to avoid taking a dependency on the FabricTransport by defining the TraceContextEnriched\*ClientFactory classes as decorators for the IServiceRemotingClientFactory interface instead of the concrete FabricTransport*ClientFactory classes?

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs
Comment (olegsych) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs:19]: Is it possible to implement TraceConstextEnriched\*RemotingDispatcher as a decorator of the IServiceRemotingMessageHandler interface instead of inheriting them two different SF classes?

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs]: Good idea. I made the change and added TraceContextEnrichedServiceRemotingClientFactoryAdapter.

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs:19]: For the dispatchers I cannot use a decorator because the base classes are different, and the methods need to be overridden.

**Comment by sablancoleis** [comment]
> One more comment, do you see any possibility to include native-instrumentation directly in ServiceFabric? Whole Azure SDK is following this pattern, and it is the best option we can recommend.

@Kielek, I got confirmation from the Service Fabric Team that they don't have plans to integrate OpenTelemetry in the near future.

File: .github/component_owners.yml
Comment (Kielek) [inline .github/component_owners.yml:61]: Is there any chance to have second volunteer to help with the maintenance?

File: .github/workflows/ci.yml
Comment (Kielek) [inline .github/workflows/ci.yml:50]: I do not see examples so this can be removed)

```suggestion
          instrumentation-servicefabricremoting: ['*/OpenTelemetry.Instrumentation.ServiceFabricRemoting*/**', '!**/*.md']
```

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj:22]: `OpenTelemetry.Api` is transitive dependency. You can simplify here:

```suggestion
    <PackageReference Include="OpenTelemetry.Api.ProviderBuilderExtensions" Version="$(OpenTelemetryCoreLatestVersion)" />
```

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md:6]: Please update

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md:19]: It is not true. I am not sure if you need an example. Or you can have cool enough README here.

**Review by Kielek — COMMENTED** [review]
Just some technical comments.
I do not know anything about ServiceFabric. In this context it will be great to find someone to review.

That is sad, the ServiceFabric does not want to implement such stuff. Maybe you can consider contributing there? I am not against including this instrumentation library here, but it will be great to find the agreement also internally within MS teams.

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md]: updated

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md]: removed

File: .github/component_owners.yml
Comment (sablancoleis) [reply .github/component_owners.yml:60]: I have been asking the Service Fabric Team for volunteers but so far no one has. I will update the owners if find somebody.

**Comment by sablancoleis** [comment]
> Just some technical comments. I do not know anything about ServiceFabric. In this context it will be great to find someone to review.
> 
> That is sad, the ServiceFabric does not want to implement such stuff. Maybe you can consider contributing there? I am not against including this instrumentation library here, but it will be great to find the agreement also internally within MS teams.

@Kielek , Oleg Sych is part of the Service Fabric Team at Microsoft and has reviewed the PR (see his comments above).

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj:9]: Analysis level is enabled for whole solution. You should remove this.

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj:8]: I think that you do not have any baseline to validation. It should be disabled as long as you do not have stable release. You can find examples in most of the instrumentation libraries how to do this.

This section

```
  <!-- Do not run Package Baseline Validation as this package has never released a stable version.
  Remove this property once we have released a stable version and add PackageValidationBaselineVersion property. -->
  <PropertyGroup>
    <DisablePackageBaselineValidation>true</DisablePackageBaselineValidation>
  </PropertyGroup>
```

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (Kielek) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj:4]: Most of the packages is providing this set of targets

```suggestion
        <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
    <TargetFrameworks>$(NetMinimumSupportedVersion);$(NetStandardMinimumSupportedVersion)</TargetFrameworks>
    <TargetFrameworks Condition="$(OS) == 'Windows_NT'">$(TargetFrameworks);$(NetFrameworkMinimumSupportedVersion)</TargetFrameworks>
```

Consider if it is not the better option for you.

**Review by Kielek — COMMENTED** [review]
@sablancoleis. sorry for delay. I have made only technical review. Some comments to be addressed. What is more, you need to fix some linters/dotnet formats. It allows as to keep the code in the common shaper.

Once again, it will be much better to include directly in instrumented library, but I will not block adding it here, as you agree in MS that it is the best option. @CodeBlanch, @rajkumar-rangaraj, @TimothyMothra, @lmolkova - it should be on your side.

**Comment by Kielek** [comment]
Note for @open-telemetry/dotnet-contrib-maintainers - remember to create label before merging.

**Review by CodeBlanch — APPROVED** [review]
I didn't review any of the implementation but the infra bits (CI, templates, etc) LGTM

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj]: fixed

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj]: fixed

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs
Comment (olegsych) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs:18]: ActorServiceRemotingDispatcher is derived from ServiceRemotingMessageDispatcher, which implements IServiceRemotingMessageHandler. Why can't a single TraceContextEnrichedRemotingDispatcher implement IServiceRemotingMessageHandler and take a constructor parameter of IServiceRemotingMessageHandler type, which could be either ActorServiceRemotingDispatcher or ServiceRemotingMessageDispatcher?

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs
Comment (olegsych) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs]: I don't think this was addressed. The factory still hardcodes the assumption that a FabricTransport is used for remoting.

Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorRemotingClientFactory.cs]: @olegsych , I think you are not looking at the latest version of the code. The new factory takes an IServiceRemotingClientFactory as a parameter in the constructor and doesn't take a dependency on FabricTransport.

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs
Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/ActorRemoting/TraceContextEnrichedActorServiceV2RemotingDispatcher.cs]: @olegsych , I have replaced the two custom dispatchers with a single Dispatcher decorator. Please take a look at the latest code. Thanks.

**Comment by Kielek** [comment]
@sablancoleis, could you please fix lint/format issues? Without this there will be not possibility to merge.

File: .github/workflows/ci.yml
Comment (Kielek) [inline .github/workflows/ci.yml:383]:
```suggestion
      project-name: Component[OpenTelemetry.Instrumentation.ServiceFabricRemoting]
```

**Comment by Kielek** [comment]
@sablancoleis, I have fixed the CI definition. To proceed, fix other build issues (file encoding and dotnet build).

To verify locally if your dotnet build pass, please execute it against `Release` instead of `Debug`

**Comment by sablancoleis** [comment]
> @sablancoleis, I think now, last blocker is your membership on OTel organization. https://github.com/open-telemetry/community/issues/new?assignees=&labels=area%2Fgithub-membership&projects=&template=membership.md&title=REQUEST%3A+New+membership+for+%3Cyour-GH-handle%3E
> 
> I can be one of your supporters.

@Kielek, Thank you for all the help with this PR. I have submitted the request.

File: .github/component_owners.yml
Comment (Kielek) [inline .github/component_owners.yml:154]:
```suggestion
  test/OpenTelemetry.Instrumentation.ServiceFabricRemoting.Tests/:
```

File: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj
Comment (sablancoleis) [inline src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj:31]: @Kielek , our applications are not ready to switch to 1.10 yet, should I change this to 1.9.0 for now?

Also, this code has been running in our production service fabric clusters for three months without any issues. Should I change the stability to RC instead of beta? We would like to start the transition to consuming this nugget as soon as it is ready.

Comment (Kielek) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj:31]: It should be fine make 1/couple of release with 1.9.0. Finally, the recommendation is to pin to the macro.

If we speaking about stability. My recommendation for the first release is either `alpha` or (as you heavily tested it) `beta`. 

Going to RC/stable phase should means that all attributes/metrics you produce are semantic version compliant (and sem conv is stable).

Comment (sablancoleis) [reply src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/OpenTelemetry.Instrumentation.ServiceFabricRemoting.csproj]: ok, sounds good. I just changed the version to 1.9.0 for now.

**Comment by sablancoleis** [comment]
@Kielek , Is there anything else holding the PR besides the unrelated build errors?

**Review by Kielek — APPROVED** [review]

**Comment by Kielek** [comment]
Technical requirements to merge to this repository are met. I was not reviewing logic behind the instrumentation.

**Comment by sablancoleis** [comment]
@Kielek how do releases work? Do I need to request a release for this package, or it will happen automatically at some point?

**Pull request merged**
